libnavigation-core/src/main/java/com/mapbox/navigation/core/trip/session/MapboxTripSession.kt

```diff
@@ -70,20 +69,24 @@ internal class MapboxTripSession(
             .build()
     }
 
-    private val mutex = Mutex()
+    private var updateNavigatorStatusDataJobs: MutableList<Job> = CopyOnWriteArrayList()
 
     override var route: DirectionsRoute? = null
         set(value) {
             field = value
+            cancelOngoingUpdateNavigatorStatusDataJobs()
             ioJobController.scope.launch {
-                mutex.withLock {
-                    navigator.setRoute(value)
-                    mainJobController.scope.launch {
-                        isOffRoute = false
-                    }
-                }
+                cancelOngoingUpdateNavigatorStatusDataJobs()
+                navigator.setRoute(value)
             }
+            isOffRoute = false
+        }
+
+    private fun cancelOngoingUpdateNavigatorStatusDataJobs() {
+        updateNavigatorStatusDataJobs.forEach {
+            it.cancel()
```

Comment:
Can each element remove itself when it's canceled? Otherwise, this list might grow to thousands of elements.

We could use a synchronized collection with an iterator instead of the `CopyOnWriteArrayList` to resolve the removal and the concurrent modification exception.

Reply:
> Can each element remove itself when it's canceled? Otherwise, this list might grow to thousands of elements.

Good point 👍 

> We could use a synchronized collection with an iterator instead of the `CopyOnWriteArrayList` to resolve the removal and the concurrent modification exception.

Tried that in the first place and run into `ConcurrentModificationException` as well 😅 https://stackoverflow.com/questions/25113987/why-is-there-a-concurrentmodificationexception-even-when-list-is-synchronized

Reply:
Oh yea, rather synchronize the access to the collection, not use a "synchronized collection", my bad :D

Reply:
> Can each element remove itself when it's canceled?

It should also remove itself when the job finishes I guess.